File: scripts/fix-prerender-assets.py
#!/usr/bin/env python3
"""
Fix Prerender Assets Script

This script ensures all prerendered HTML files have the correct production
CSS and JS asset references. Run this after building to fix any missing assets.

Usage: python3 scripts/fix-prerender-assets.py
"""

import os
import re
import sys

def main():
    prerender_dir = 'dist/prerendered'
    prod_index = 'dist/public/index.html'
    
    print("🔧 Fixing Prerender Assets\n")
    print("━" * 50)
    
    # Check directories
    if not os.path.exists(prerender_dir):
        print(f"❌ Prerender directory not found: {prerender_dir}")
        sys.exit(1)
    if not os.path.exists(prod_index):
        print(f"❌ Production index.html not found: {prod_index}")
        sys.exit(1)
    
    # Read production HTML
    with open(prod_index, 'r') as f:
        prod_html = f.read()
    
    # Extract production assets
    css_match = re.search(r'<link rel="stylesheet" crossorigin href="/assets/index[^"]*\.css">', prod_html)
    js_match = re.search(r'<script type="module" crossorigin src="/assets/index[^"]*\.js"></script>', prod_html)
    
    prod_css = css_match.group() if css_match else ''
    prod_js = js_match.group() if js_match else ''
    
    print(f"📦 CSS: {prod_css[:60]}...")
    print(f"📦 JS:  {prod_js[:60]}...\n")
    
    if not prod_css or not prod_js:
        print("❌ Could not extract production assets!")
        sys.exit(1)
    
    # Process all HTML files
    fixed_css = 0
    fixed_js = 0
    total_files = 0
    
    for root, dirs, files in os.walk(prerender_dir):
        for fname in files:
            if not fname.endswith('.html'):
                continue
                
            total_files += 1
            fpath = os.path.join(root, fname)
            
            try:
                with open(fpath, 'r') as f:
                    html = f.read()
                
                original_html = html
                
                # Remove old/dev asset references
                html = re.sub(r'<script[^>]*src="/src/main\.tsx[^"]*"[^>]*></script>', '', html)
                html = re.sub(r'<link[^>]*rel="modulepreload"[^>]*href="/src/[^"]*"[^>]*>', '', html)
                html = re.sub(r'<script[^>]*src="/@replit/[^"]*"[^>]*></script>', '', html)
                html = re.sub(r'<script[^>]*src="/@vite/[^"]*"[^>]*></script>', '', html)
                
                # Add CSS if missing
                if prod_css not in html:
                    if '</head>' in html:
                        html = html.replace('</head>', f'    {prod_css}\n  </head>')
                        fixed_css += 1
                
                # Add JS if missing
                if prod_js not in html:
                    if '</body>' in html:
                        html = html.replace('</body>', f'    {prod_js}\n  </body>')
                        fixed_js += 1
                
                # Write if changed
                if html != original_html:
                    with open(fpath, 'w') as f:
                        f.write(html)
                        
            except Exception as e:
                print(f"⚠️  Error processing {fpath}: {e}")
    
    print("━" * 50)
    print("📊 SUMMARY\n")
    print(f"Total files:     {total_files}")
    print(f"Fixed CSS:       {fixed_css}")
    print(f"Fixed JS:        {fixed_js}")
    print()
    
    if fixed_css == 0 and fixed_js == 0:
        print("✅ All files already have correct assets")
    else:
        print(f"✅ Fixed {fixed_css + fixed_js} asset references")
    
    print("\n🔍 Run validation: npx tsx scripts/validate-prerender.ts")

if __name__ == '__main__':
    main()

File: scripts/fix-prerender-assets.ts
#!/usr/bin/env tsx
/**
 * Fix Prerender Assets Script
 *
 * This script ensures all prerendered HTML files have the correct production
 * CSS and JS asset references. Run this after building to fix any missing assets.
 *
 * Usage: npx tsx scripts/fix-prerender-assets.ts
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const rootDir = path.resolve(__dirname, '..');

const PRERENDER_DIR = path.join(rootDir, 'dist/prerendered');
const PROD_INDEX = path.join(rootDir, 'dist/public/index.html');

function walkDir(dir: string): string[] {
  const files: string[] = [];
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  
  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      files.push(...walkDir(fullPath));
    } else if (entry.name.endsWith('.html')) {
      files.push(fullPath);
    }
  }
  
  return files;
}

function main() {
  console.log('🔧 Fixing Prerender Assets\n');
  console.log('━'.repeat(50));

  // Check directories exist
  if (!fs.existsSync(PRERENDER_DIR)) {
    console.error(`❌ Prerender directory not found: ${PRERENDER_DIR}`);
    process.exit(1);
  }
  if (!fs.existsSync(PROD_INDEX)) {
    console.error(`❌ Production index.html not found: ${PROD_INDEX}`);
    process.exit(1);
  }

  // Read production HTML
  const prodHtml = fs.readFileSync(PROD_INDEX, 'utf-8');

  // Extract production assets
  const cssMatch = prodHtml.match(/<link rel="stylesheet" crossorigin href="\/assets\/index[^"]*\.css">/);
  const jsMatch = prodHtml.match(/<script type="module" crossorigin src="\/assets\/index[^"]*\.js"><\/script>/);

  const prodCss = cssMatch ? cssMatch[0] : '';
  const prodJs = jsMatch ? jsMatch[0] : '';

  console.log(`📦 CSS: ${prodCss.substring(0, 60)}...`);
  console.log(`📦 JS:  ${prodJs.substring(0, 60)}...\n`);

  if (!prodCss || !prodJs) {
    console.error('❌ Could not extract production assets!');
    process.exit(1);
  }

  // Process all HTML files
  let fixedCss = 0;
  let fixedJs = 0;
  let totalFiles = 0;

  const htmlFiles = walkDir(PRERENDER_DIR);

  for (const fpath of htmlFiles) {
    totalFiles++;

    try {
      let html = fs.readFileSync(fpath, 'utf-8');
      const originalHtml = html;

      // Remove old/dev asset references
      html = html.replace(/<script[^>]*src="\/src\/main\.tsx[^"]*"[^>]*><\/script>/g, '');
      html = html.replace(/<link[^>]*rel="modulepreload"[^>]*href="\/src\/[^"]*"[^>]*>/g, '');
      html = html.replace(/<script[^>]*src="\/@replit\/[^"]*"[^>]*><\/script>/g, '');
      html = html.replace(/<script[^>]*src="\/@vite\/[^"]*"[^>]*><\/script>/g, '');

      // Add CSS if missing
      if (!html.includes(prodCss)) {
        if (html.includes('</head>')) {
          html = html.replace('</head>', `    ${prodCss}\n  </head>`);
          fixedCss++;
        }
      }

      // Add JS if missing
      if (!html.includes(prodJs)) {
        if (html.includes('</body>')) {
          html = html.replace('</body>', `    ${prodJs}\n  </body>`);
          fixedJs++;
        }
      }

      // Write if changed
      if (html !== originalHtml) {
        fs.writeFileSync(fpath, html, 'utf-8');
      }
    } catch (e: any) {
      console.error(`⚠️  Error processing ${fpath}: ${e.message}`);
    }
  }

  console.log('━'.repeat(50));
  console.log('📊 SUMMARY\n');
  console.log(`Total files:     ${totalFiles}`);
  console.log(`Fixed CSS:       ${fixedCss}`);
  console.log(`Fixed JS:        ${fixedJs}`);
  console.log();

  if (fixedCss === 0 && fixedJs === 0) {
    console.log('✅ All files already have correct assets');
  } else {
    console.log(`✅ Fixed ${fixedCss + fixedJs} asset references`);
  }

  console.log('\n🔍 Run validation: npx tsx scripts/validate-prerender.ts');
}

main();
